source/common/shared_pool/shared_pool.h

```diff
@@ -30,60 +31,63 @@ namespace SharedPool {
  * There is also a need to ensure that the thread where ObjectSharedPool's destructor is also in the
  * main thread, or that ObjectSharedPool destruct before the program exit
  */
-template <typename T, typename HashFunc = std::hash<T>,
+template <typename T, typename HashFunc = std::hash<T>, typename EqualFunc = std::equal_to<T>,
           class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
-class ObjectSharedPool : public Singleton::Instance,
-                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
-                         NonCopyable {
+class ObjectSharedPool
+    : public Singleton::Instance,
+      public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc, EqualFunc>>,
+      NonCopyable {
 public:
   ObjectSharedPool(Event::Dispatcher& dispatcher)
       : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
 
-  void deleteObject(const size_t hash_key) {
+  void deleteObject(T* ptr) {
     if (std::this_thread::get_id() == thread_id_) {
-      // There may be new inserts with the same hash value before deleting the old element,
-      // so there is no need to delete it at this time.
-      if (object_pool_.find(hash_key) != object_pool_.end() &&
-          object_pool_[hash_key].use_count() == 0) {
-        object_pool_.erase(hash_key);
+      if (auto iter = object_pool_.find(ptr); iter != object_pool_.end()) {
```

Comment:
my high level thought is that this whole class would be a lot simpler if we just protected the map with a mutex, which would be uncontended most of the time. Was there ever any data showing perf bottlenecks due to the mutex?

This might be aa question for the original writer @zyfjeff and reviewer(s) @htuch (IIRC).